File: web/server/routes/memos.ts
import { Hono } from 'hono';
import { getDb, getWriteDb } from '../db.js';

const app = new Hono();

// GET /api/memos?date=YYYY-MM-DD
app.get('/', (c) => {
  const date = c.req.query('date');
  if (!date) return c.json({ error: 'date required' }, 400);

  const db = getDb();
  const row = db
    .prepare('SELECT content, updated_at FROM memos WHERE date = ?')
    .get(date) as { content: string; updated_at: string } | undefined;

  return c.json({ date, content: row?.content ?? '', updated_at: row?.updated_at ?? null });
});

// PUT /api/memos — upsert memo (today only)
app.put('/', async (c) => {
  const body = await c.req.json<{ date: string; content: string }>();
  if (!body.date || typeof body.content !== 'string') {
    return c.json({ error: 'date and content required' }, 400);
  }

  // Only allow editing today's memo
  const today = new Date().toISOString().slice(0, 10);
  if (body.date !== today) {
    return c.json({ error: 'can only edit today\'s memo' }, 403);
  }

  const db = getWriteDb();
  db.prepare(
    `INSERT INTO memos (date, content, updated_at) VALUES (?, ?, datetime('now'))
     ON CONFLICT(date) DO UPDATE SET content=excluded.content, updated_at=datetime('now')`
  ).run(body.date, body.content);

  return c.json({ ok: true });
});

export default app;

File: web/src-tauri/src/commands/memos.rs
use crate::db::AppDb;
use crate::models::Memo;
use chrono::Local;
use tauri::State;

#[tauri::command]
pub fn get_memo(date: String, db: State<AppDb>) -> Result<Option<Memo>, String> {
    crate::commands::validate::validate_date(&date)?;
    let conn = db.conn.lock().map_err(|e| e.to_string())?;

    let mut stmt = conn
        .prepare("SELECT date, content, updated_at FROM memos WHERE date = ?1")
        .map_err(|e| e.to_string())?;

    let mut rows = stmt
        .query_map(rusqlite::params![date], |row| {
            Ok(Memo {
                date: row.get::<_, Option<String>>(0)?.unwrap_or_default(),
                content: row.get::<_, Option<String>>(1)?.unwrap_or_default(),
                updated_at: row.get::<_, Option<String>>(2)?,
            })
        })
        .map_err(|e| e.to_string())?;

    match rows.next() {
        Some(row) => Ok(Some(row.map_err(|e| e.to_string())?)),
        None => Ok(Some(Memo {
            date,
            content: String::new(),
            updated_at: None,
        })),
    }
}

#[tauri::command]
pub fn put_memo(date: String, content: String, db: State<AppDb>) -> Result<Memo, String> {
    crate::commands::validate::validate_date(&date)?;
    if content.len() > 65536 {
        return Err("memo too large".to_string());
    }
    // Only allow editing today's memo (use local date, not UTC)
    let today = Local::now().format("%Y-%m-%d").to_string();
    if date != today {
        return Err("can only edit today's memo".to_string());
    }

    let conn = db.conn.lock().map_err(|e| e.to_string())?;

    conn.execute(
        "INSERT INTO memos (date, content, updated_at) VALUES (?1, ?2, datetime('now'))
         ON CONFLICT(date) DO UPDATE SET content=excluded.content, updated_at=datetime('now')",
        rusqlite::params![date, content],
    )
    .map_err(|e| e.to_string())?;

    // Read back the updated_at
    let updated_at: Option<String> = conn
        .query_row(
            "SELECT updated_at FROM memos WHERE date = ?1",
            rusqlite::params![date],
            |row| row.get(0),
        )
        .unwrap_or(None);

    Ok(Memo {
        date,
        content,
        updated_at,
    })
}
